File: 054__Spiral_Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if len(matrix) == 0:
            return []

        m = len(matrix)
        n = len(matrix[0])

        visited = [[False] * n for _ in range(m)]
        visited_num = 1
        result = [matrix[0][0]]
        visited[0][0] = True
        direct = [0, 1]
        i, j = 0, 0
        while visited_num < m * n:
            if direct == [0, 1]:
                while j + 1 < n and not visited[i][j + 1] :
                    visited[i][j + 1] = True
                    j += 1
                    visited_num += 1
                    result += [matrix[i][j]]
                direct = [1, 0]
            elif direct == [1, 0]:
                while i + 1 < m and not visited[i + 1][j] :
                    visited[i + 1][j] = True
                    i += 1
                    visited_num += 1
                    result += [matrix[i][j]]
                direct = [0, -1]
            elif direct == [0, -1]:
                while j - 1 >= 0 and not visited[i][j - 1]:
                    visited[i][j - 1] = True
                    j -= 1
                    visited_num += 1
                    result += [matrix[i][j]]
                direct = [-1, 0]
            else:
                while i - 1 >= 0 and not visited[i - 1][j]:
                    visited[i - 1][j] = True
                    i -= 1
                    visited_num += 1
                    result += [matrix[i][j]]
                direct = [0, 1]
        return result

        

File: Java/054__Spiral_Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List result = new ArrayList();
        if (matrix.length == 0 || matrix[0].length == 0) {
            return result;
        }
        int r_start = 0, r_end = matrix.length - 1;
        int c_start = 0, c_end = matrix[0].length - 1;
        while (r_start <= r_end && c_start <= c_end) {
            for (int j = c_start; j <= c_end; j++) {
                result.add(matrix[r_start][j]);
            }
            for (int i = r_start + 1; i <= r_end; i++) {
                result.add(matrix[i][c_end]);
            }
            if (r_start < r_end && c_start < c_end) {
                for (int j = c_end - 1; j >= c_start; j--) {
                    result.add(matrix[r_end][j]);
                }
                for (int i = r_end - 1; i >= r_start + 1; i--) {
                    result.add(matrix[i][c_start]);
                }
            }
            r_start++;
            r_end--;
            c_start++;
            c_end--;
        }
        return result;
    }
}
